ext/LinearSolveHYPRE.jl

```diff
@@ -0,0 +1,219 @@
+    end
+
+    # Otherwise instantiate
+    if !(alg.solver <: Union{COMM_SOLVERS, NO_COMM_SOLVERS})
+        throw(ArgumentError("unknown or unsupported HYPRE solver: $(alg.solver)"))
+    end
+    comm = cache.cacheval.A.comm # communicator from the matrix
+    solver = create_solver(alg.solver, comm)
+
+    # Construct solver options
+    solver_options = (;
+                      AbsoluteTol = cache.abstol,
+                      MaxIter = cache.maxiters,
+                      PrintLevel = Int(cache.verbose),
+                      Tol = cache.reltol)
+
+    # Preconditioner (uses Pl even though it might not be a *left* preconditioner just *a*
+    # preconditioner)
+    if !(cache.Pl isa Identity)
+        P = if cache.Pl isa HYPRESolver
+            cache.Pl
+        elseif cache.Pl <: HYPRESolver
+            create_solver(cache.Pl, comm)
+        else
+            throw(ArgumentError("unknown HYPRE preconditioner $(cache.Pl)"))
+        end
+        solver_options = merge(solver_options, (; Precond = P))
+    end
+
+    # Filter out some options that are not supported for some solvers
+    if solver isa HYPRE.Hybrid
+        # Rename MaxIter to PCGMaxIter
+        MaxIter = solver_options.MaxIter
+        ks = filter(x -> x !== :MaxIter, keys(solver_options))
+        solver_options = NamedTuple{ks}(solver_options)
+        solver_options = merge(solver_options, (; PCGMaxIter = MaxIter))
+    elseif solver isa HYPRE.BoomerAMG || solver isa HYPRE.ILU
+        # Remove AbsoluteTol, Precond
+        ks = filter(x -> !in(x, (:AbsoluteTol, :Precond)), keys(solver_options))
+        solver_options = NamedTuple{ks}(solver_options)
+    end
+
+    # Set the options
+    HYPRE.Internals.set_options(solver, pairs(solver_options))
+
+    return solver
+end
+
+# TODO: How are args... and kwargs... supposed to be used here?
+function SciMLBase.solve(cache::LinearCache, alg::HYPREAlgorithm, args...; kwargs...)
+    # It is possible to reach here without HYPRE.Init() being called if HYPRE structures are
+    # only to be created here internally (i.e. when cache.A::SparseMatrixCSC and not a
+    # ::HYPREMatrix created externally by the user). Be nice to the user and call it :)
+    if !(cache.A isa HYPREMatrix || cache.b isa HYPREVector || cache.u isa HYPREVector ||
+         alg.solver isa HYPRESolver)
+        HYPRE.Init()
+    end
+
+    # Move matrix and vectors to HYPRE, if not already provided as HYPREArrays
+    hcache = cache.cacheval
+    if hcache.isfresh_A || hcache.A === nothing
+        hcache.A = cache.A isa HYPREMatrix ? cache.A : HYPREMatrix(cache.A)
+        hcache.isfresh_A = false
+    end
+    if hcache.isfresh_b || hcache.b === nothing
+        hcache.b = cache.b isa HYPREVector ? cache.b : HYPREVector(cache.b)
+        hcache.isfresh_b = false
+    end
+    if hcache.isfresh_u || hcache.u === nothing
+        hcache.u = cache.u isa HYPREVector ? cache.u : HYPREVector(cache.u)
+        hcache.isfresh_u = false
+    end
```

Comment:
Does this play nicely with Distributed? So for example, if someone used a `DistributedArray` `A` and `b`, will this nicely use the memory? Copy it again? Distribute the same way? I'm curious because if we could have a default solver choice on DistributedArray values then that will fix a few downstream issues.

Reply:
Right now it will only work if you have either set up your matrix and vector with HYPRE to begin with or if you use a `SparseMatrixCSC` or `SparseMatrixCSR`. For the latter the matrix will be duplicated inside the hypre library, and some buffers need to be allocated in order to send the data to HYPRE.

I don't think it would work with Distributed or DistributedArray since those are not MPI processes, right?

Reply:
Makes sense. With Distributed you could make it use the MPI cluster manager, but that's not too common.